### PR TITLE
[Expression] Make color functions compatible with color object

### PR DESCRIPTION
**Pull request by troopa81**
... and not only color string representation as it as been discussed in https://github.com/qgis/QGIS/pull/58458#issuecomment-2303538480

Concerned functions:
- color_grayscale_average
- color_mix
- set_color_part
- color_part
- darker
- lighter

**Funded by Bordeaux Métropôle**

### CODE REVIEW

File: src/core/expression/qgsexpressionfunction.cpp [hunk 1]
```diff
@@ -5821,13 +5821,23 @@ static QVariant fcnFormatDate( const QVariantList &values, const QgsExpressionCo
 
 static QVariant fcnColorGrayscaleAverage( const QVariantList &values, const QgsExpressionContext *, QgsExpression *, const QgsExpressionNodeFunction * )
 {
-  QColor color = QgsSymbolLayerUtils::decodeColor( values.at( 0 ).toString() );
-  int avg = ( color.red() + color.green() + color.blue() ) / 3;
-  int alpha = color.alpha();
+  const QVariant variant = values.at( 0 );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp:5824]: This logic should be moved to QgsExpressionUtils, as getColorValue, with isQColor as a by reference argument

Comment (troopa81) [reply src/core/expression/qgsexpressionfunction.cpp:5824]: Done 06048538d4d

File: tests/src/core/testqgsexpression.cpp [hunk 2]
```diff
@@ -2314,33 +2331,33 @@ class TestQgsExpression: public QObject
           switch ( resultColor.spec() )
           {
             case QColor::Spec::Cmyk:
-              QVERIFY( qgsDoubleNear( resultColor.cyanF(), expectedColor.cyanF(), 0.001 ) );
-              QVERIFY( qgsDoubleNear( resultColor.magentaF(), expectedColor.magentaF(), 0.001 ) );
-              QVERIFY( qgsDoubleNear( resultColor.yellowF(), expectedColor.yellowF(), 0.001 ) );
-              QVERIFY( qgsDoubleNear( resultColor.blackF(), expectedColor.blackF(), 0.001 ) );
-              QVERIFY( qgsDoubleNear( resultColor.alphaF(), expectedColor.alphaF(), 0.001 ) );
+              QVERIFY2( qgsDoubleNear( resultColor.cyanF(), expectedColor.cyanF(), 0.001 ), QString( "cyan: %2!=%3" ).arg( resultColor.cyanF(), 0, 'f', 3 ).arg( expectedColor.cyanF(), 0, 'f', 3 ).toLatin1() );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp:2334]: Couldn't we use QGSCOMPARENEAR instead?

Comment (troopa81) [reply tests/src/core/testqgsexpression.cpp]: Yes! Done [a55451b](https://github.com/qgis/QGIS/pull/58561/commits/a55451b6fc993f65df1f03d46a0b651f172d2a64)

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**